src/iceberg_destination.rs

```diff
@@ -0,0 +1,201 @@
+};
+use iceberg::writer::file_writer::{FileWriter, FileWriterBuilder, ParquetWriterBuilder};
+use iceberg::TableCreation;
+use log::info;
+use parquet::arrow::PARQUET_FIELD_ID_META_KEY;
+use parquet::file::properties::WriterProperties;
+use url::Url;
+use uuid::Uuid;
+
+use crate::error::DataLoadingError;
+
+pub async fn record_batches_to_iceberg(
+    record_batch_stream: impl TryStream<Item = Result<RecordBatch, DataLoadingError>>,
+    arrow_schema: SchemaRef,
+    target_url: Url,
+) -> Result<(), DataLoadingError> {
+    pin_mut!(record_batch_stream);
+
+    let mut file_io_props: Vec<(String, String)> = vec![];
+    if let Ok(aws_endpoint) = std::env::var("AWS_ENDPOINT") {
+        file_io_props.push(("s3.endpoint".to_string(), aws_endpoint));
+    }
+
+    let file_io = FileIO::from_path(target_url.clone())?
+        .with_props(file_io_props)
+        .build()?;
+
+    // Assign sequential field IDs starting from 1 to the fields
+    let mut field_id_counter = 1;
+    let fields: Vec<Field> = arrow_schema
+        .fields
+        .iter()
+        .map(|field_ref| {
+            let mut field: Field = (**field_ref).clone();
+            let mut metadata = field_ref.metadata().clone();
+            metadata.insert(
+                PARQUET_FIELD_ID_META_KEY.to_owned(),
+                field_id_counter.to_string(),
+            );
+            field_id_counter += 1;
+            field.set_metadata(metadata);
+            field
+        })
+        .collect();
+
+    let cloned_arrow_schema = Schema::new_with_metadata(fields, arrow_schema.metadata.clone());
+
+    let iceberg_schema = iceberg::arrow::arrow_schema_to_schema(&cloned_arrow_schema)?;
+
+    let table_creation = TableCreation::builder()
+        .name("dummy_name".to_string()) // Required by TableCreationBuilder. Doesn't affect output
+        .schema(iceberg_schema.clone())
+        .location(target_url.to_string())
+        .build();
+
+    let metadata_v0 = TableMetadataBuilder::from_table_creation(table_creation)?.build()?;
+    let metadata_v0_location = format!("{}/metadata/v0.metadata.json", target_url);
+
+    file_io
+        .new_output(&metadata_v0_location)?
+        .write(serde_json::to_vec(&metadata_v0).unwrap().into())
+        .await?;
+    info!("Wrote v0 metadata: {:?}", metadata_v0_location);
+
+    let file_writer_builder = ParquetWriterBuilder::new(
+        WriterProperties::builder().build(),
+        Arc::new(iceberg_schema.clone()),
+        file_io.clone(),
+        DefaultLocationGenerator::new(metadata_v0.clone()).unwrap(),
+        DefaultFileNameGenerator::new(
+            "part".to_string(),
+            Some(Uuid::new_v4().to_string()),
+            DataFileFormat::Parquet,
+        ),
+    );
+    let mut file_writer = file_writer_builder.build().await.unwrap();
+
+    while let Some(maybe_batch) = record_batch_stream.next().await {
+        let batch = maybe_batch?;
+        file_writer.write(&batch).await?;
```

Comment:
I think this results in one or more row groups being buffered in memory prior to flushing, which may (or may not) pose memory usage issues. If it turns out it does we'd need to customize it, like we do with delta (temp files + multipart upload), but for start I think it's good.